src/OpenTelemetry.Exporter.Geneva/GenevaLogExporter.cs

```diff
@@ -418,6 +418,44 @@ internal int SerializeLogRecord(LogRecord logRecord)
             cntFields += 1;
         }
 
+        ushort scopeDepth = 0;
+        int indexArrayLength = 0;
+        logRecord.ForEachScope(ProcessScope, (object)null);
```

Comment:
FYI this is going to allocate a delegate. Could maybe get around that using a `ThreadStatic`. Something like this...

```csharp
private sealed class ScopeForEachState
{
   public ushort ScopeDepth;
   public int IndexArrayLength; // Confusing name
   public byte[] Buffer;
}

[ThreadStatic]
private static ScopeForEachState scopeState;

var state = scopeState ??= new() { Buffer = buffer };

state.ScopeDepth = 0;
state.IndexArrayLength = 0;

logRecord.ForEachScope(ProcessScope, state);

if (state.ScopeDepth > 0)
{
        MessagePackSerializer.WriteUInt16(buffer, state.IndexArrayLength, state.ScopeDepth);
        cntFields += 1;
}

private static readonly Action<LogRecordScope, ScopeForEachState> ProcessScope = (logRecord, state) =>
{
           var buffer = state.Buffer;

            if (++state.ScopeDepth == 1)
            {
                cursor = MessagePackSerializer.SerializeAsciiString(buffer, cursor, "scopes");
                cursor = MessagePackSerializer.WriteArrayHeader(buffer, cursor, ushort.MaxValue);
                state.IndexArrayLength = cursor - 2;
            }

            cursor = MessagePackSerializer.WriteMapHeader(buffer, cursor, ushort.MaxValue);
            int indexMapSizeScope = cursor - 2;
            ushort keysCount = 0;

            foreach (KeyValuePair<string, object> scopeItem in scope)
            {
                string key = "scope";
                if (!string.IsNullOrEmpty(scopeItem.Key))
                {
                    key = scopeItem.Key;
                }

                cursor = MessagePackSerializer.SerializeUnicodeString(buffer, cursor, key);
                cursor = MessagePackSerializer.Serialize(buffer, cursor, scopeItem.Value);
                keysCount++;
            }

            MessagePackSerializer.WriteUInt16(buffer, indexMapSizeScope, keysCount);
};
```

Reply:
We have `private static readonly ThreadLocal<byte[]> m_buffer = new ThreadLocal<byte[]>(() => null);` you could just expand that a little bit to be like `ThreadLocal<ExporterState>` and then use `ExporterState` to contain the buffer + the stuff you need for scopes. Basically fold everything you need onto the 1 `ThreadLocal`.

Reply:
Thanks @CodeBlanch ! I have created #391 to track this. I'll be merging this PR to do a beta release to gather feedback. I'll send out a separate PR to avoid allocation here.